# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [2]:
import numpy as np

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
data.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [7]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [8]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [9]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [10]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [11]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [12]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Сгруппируем данные по количеству детей и посмотрим на среднее значение по столбу debt, так же выведем количество клиентов и сколько из них должников

In [24]:
data.pivot_table(index='children',values='debt',aggfunc=['count','sum','mean'])

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

Сгруппируем данные, выделим в новый столбец - есть дети / нет детей, и повторим  группировку.

In [25]:
data['children_have'] = (data['children'] > 0) * 1

In [26]:
data.pivot_table(index='children_have',values='debt',aggfunc=['count','sum','mean'])

count   sum      mean
                debt  debt      debt
children_have                       
0              14091  1063  0.075438
1               7240   669  0.092403

**Вывод:** Можно заметить, что люди без детей более ответсвенные заемщики, но нельзя утверждать что именно наличие или отсутствие детей влияет на результат. Посмотрим, возможно далее будет какая то более явная закономерность.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Сгруппируем данные по семейному положению и посмотрим на среднее значение по столбу debt, для удобства отсортируем от меньшего к большему

In [27]:
data.groupby(['family_status'])['debt'].agg(['count','mean']).sort_values(by = 'mean' )

count      mean
family_status                         
вдовец / вдова           951  0.066246
в разводе               1189  0.070648
женат / замужем        12261  0.075606
гражданский брак        4134  0.093130
Не женат / не замужем   2796  0.097639

Выводы немного удивили, добавлю сколько из них должников, возможно это даст больше ответов.

In [28]:
data.groupby(['family_status'])['debt'].agg(['count','sum','mean']).sort_values(by = 'mean' )

count  sum      mean
family_status                              
вдовец / вдова           951   63  0.066246
в разводе               1189   84  0.070648
женат / замужем        12261  927  0.075606
гражданский брак        4134  385  0.093130
Не женат / не замужем   2796  273  0.097639

**Вывод:** Лучше не стало, хуже всех показатель у холостых, это ожидаемо. А вот группа женат/замужем держатся в середине и почти на 1% обгоняют вдовец / вдова. Для меня вдовец / вдова ни чем не отличается от не женат / не замужем. Да и в разводе, тот же холостой. Правильно наверно будет рассмотреть семейное положение как две группы был в браке или нет, тогда можно сделать вывод, что клиенты находящиеся/находившееся в браке боле надежные заемщики

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Сгруппируем данные по доходу и посмотрим на среднее значение по столбу debt

In [29]:
data.pivot_table(index='total_income_category',values='debt',aggfunc=['count','sum','mean'])

count   sum      mean
                        debt  debt      debt
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

Самая большая выборка получилась в категории С, для более детального изучения, разобьем на 7 категорий

In [30]:
data['total_income_group'] = pd.cut(data['total_income'],[0,50000,100000,150000,200000,250000,300000,np.inf])

In [31]:
data.pivot_table(index='total_income_group',values='debt',aggfunc=['count','sum','mean'])

count  sum      mean
                      debt debt      debt
total_income_group                       
(0.0, 50000.0]         371   23  0.061995
(50000.0, 100000.0]   4073  331  0.081267
(100000.0, 150000.0]  7110  619  0.087060
(150000.0, 200000.0]  4738  403  0.085057
(200000.0, 250000.0]  2242  162  0.072257
(250000.0, 300000.0]  1323   88  0.066515
(300000.0, inf]       1474  106  0.071913

**Вывод:** Более менее процент возврата у клиентов с доходом 250000-300000, так же не плохо себя показывают люди с доходом до 50000, но там маленькая группа, всего 371 человек. Хуже всего выборка от 50000 до 200000. При дополнительной разбивки показатель у этой группы держится на уровне 8-9%, Не стоит забывать что порядка 10% датафрейма мы заполнили медианным значением, не совсем корректно опираться на полученный результат

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Повторим уже стандартную операцию, сгруппируем по целевому назначению кредита, выведем количество и среднее. Всего 4 группы, сортировать не будем, так посмотрим.

In [32]:
data.pivot_table(index='purpose_category',values='debt',aggfunc=['count','sum','mean'])

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

**Вывод:** Покупка автомобиля, самый большой процент 'просрочки'. Объяснений может быть много, разбил машину, нет желания платит или покупал под такси, разбил и пока в ремонте нет денег на платеж. Может быть это была просто спонтанная покупка и теперь нет понимания как расплатиться. На свадьбу думаю часто берут или кредит или в долг с планами закрыть 'подарками'. Недвижимость это серьезная покупка и скорее всего крупная сумма, ее совершают обдуманно. Так же почти 10% 'просрочки' это образование, тут возможно клиент так же совершает спонтанную покупку, а потом перегорает.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Одна из причин, это может быть техническая ошибка или возможно в анкете эти графы не столь заметны или необязательны к заполнению. Так же возможно клиенты у кого маленькая зп или кто мало дней работает, думает, что пропуск увеличивает шанс на получение кредита, тем самым оставляют эти графы пустыми.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Что бы ответить на данный вопрос, надо посмотреть среднее и медиану, правильно было сделать до заполнения, но посмотрим тут, добавим так же max, это поможет понять , почему среднее выше медианы

In [33]:
data.groupby(['income_type'])['total_income'].agg(['count','mean', 'median','max'])

count           mean  median      max
income_type                                           
безработный          2  131339.000000  131339   202722
в декрете            1   53829.000000   53829    53829
госслужащий       1451  168728.454859  150447   910451
компаньон         5047  199608.615811  172357  2265604
пенсионер         3812  135320.753673  118514   735103
предприниматель      2  499163.000000  499163   499163
сотрудник        11015  159547.951975  142594  1726276
студент              1   98201.000000   98201    98201

*Ответ:* Мы видим, что в группах, где более менее большая выборка, значения среднее выше медианы, т.к. есть клиенты с высоким доходом. Так же из сводной таблицы можно заметить , что средне и медиана сильно отличаются от типа занятости. Поэтому корректно заполнить доход медианным значением и обязательно по  типу занятости

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

Перед тем как сделать общий вывод, думаю стоит посмотреть на возрастную группу клиентов и возможные 'просрочки' по платежам

In [40]:
data['dob_years_group'] = pd.qcut(data['dob_years'],7)

In [41]:
data.groupby(['dob_years_group'])['debt'].agg(['count','mean']).sort_values(by = 'mean' )

count      mean
dob_years_group                 
(58.0, 75.0]      2931  0.052201
(52.0, 58.0]      2756  0.063498
(45.0, 52.0]      3417  0.069359
(40.0, 45.0]      2735  0.078611
(35.0, 40.0]      2848  0.087430
(29.0, 35.0]      3378  0.102427
(-0.001, 29.0]    3266  0.109308

**Вывод** В данном исследовании мы рассмотрели группу клиентов, посмотрели как возраст, семейное положение, наличии детей и цель кредита влияет на возврат долга. Из проведеного анализа можно выделить группу, более надежных клиентов:
* Люди без детей 
* Клиенты с доходом 250-300т
* Возраст заемщика. Чем взрослее клиент, тем он надежнее
* Клиенты в браке или который состояли в браке более более ответственные заемщики 

Так же выделим 'группу риска' по целевому назначению кредита:
* покупка автомобиля
* получение образования

#### решение через функцию

In [44]:
def grouping(category):
    return data.groupby(category)['debt'].mean().sort_values()

In [45]:
grouping('children_have')

children_have
0    0.075438
1    0.092403
Name: debt, dtype: float64

In [46]:
def groupings(category):
    return data.groupby(category)['debt'].mean().to_frame().sort_values(by='debt')

In [47]:
groupings('purpose_category')

debt
purpose_category                  
операции с недвижимостью  0.072551
проведение свадьбы        0.079118
получение образования     0.092528
операции с автомобилем    0.093480

In [48]:
def grouping_ss(category):
    return data.groupby([category])['debt'].agg(['count','sum','mean']).sort_values(by = 'mean' )

In [49]:
grouping_ss('family_status')

count  sum      mean
family_status                              
вдовец / вдова           951   63  0.066246
в разводе               1189   84  0.070648
женат / замужем        12261  927  0.075606
гражданский брак        4134  385  0.093130
Не женат / не замужем   2796  273  0.097639

### тестовая таблица

In [50]:
np.nan

nan

In [60]:
test = pd.DataFrame({'x':['a','b','c','d'], 'y':[1, 2, 3, 4], 'z':['e','f','g',np.nan], 'w':['5','f','','g']})

In [61]:
test

x  y    z  w
0  a  1    e  5
1  b  2    f  f
2  c  3    g   
3  d  4  NaN  g

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
В целом, работа хорошая. У тебя отличные аналитические навыки, но остались небольшие доработки. Доделывай и присылай на повторную проверку.
        
Жду твоих исправлений 😊
</font></b>
</div>